client/service/src/task_manager.rs

```diff
@@ -281,9 +281,8 @@ impl TaskManager {
 	}
 
 	/// Move a struct into the task manager to be kept alive.
-	pub fn keep_alive<T: 'static + Send>(&mut self, to_keep_alive: T) {
-		let keeping_alive = std::mem::replace(&mut self.keep_alive, Box::new(()));
-		self.keep_alive = Box::new((keeping_alive, to_keep_alive));
+	pub fn keep_alive<T: 'static + Send + Sync>(&mut self, to_keep_alive: T) {
```

Comment:
You can change the visibility here to pub(crate) as this is implementation detail now

Reply:
https://github.com/paritytech/substrate/pull/6352/files#diff-054cba4fa464ae867c597bc0ede3fb18R327 we're still using it here to keep the rpc handlers alive.

Reply:
I think the only place it is used externally (to sc-service) is here: https://github.com/paritytech/substrate/blob/ashley-service-multiple-return/bin/node/cli/src/browser.rs#L60

It might be worth moving it to the task_manager as a separate field.

:thinking: looks like the task manager has become our new "service", our node instance.

Reply:
Everywhere it looks like it's passed around:
```
[0] [14:44:18] ~/r/substrate ashley-service-multiple-return > rg rpc_handlers
utils/browser/src/lib.rs
119:pub fn start_client(mut task_manager: TaskManager, rpc_handlers: RpcHandlers) -> Client {
130:                                    let fut = rpc_handlers

client/service/src/builder.rs
1098:           let rpc_handlers = RpcHandlers(gen_handler(sc_rpc::DenyUnsafe::No));
1143:                   rpc_handlers,

client/service/src/lib.rs
169:    pub rpc_handlers: RpcHandlers,

bin/node/cli/src/browser.rs
55:     let (task_manager, rpc_handlers) =
57:                     .map(|(components, rpc_handlers, _, _, _)| (components, rpc_handlers))
60:     Ok(browser_utils::start_client(task_manager, rpc_handlers))

bin/node/cli/src/chain_spec.rs
434:                            let (keep_alive, _, rpc_handlers, client, network, transaction_pool) = new_full_base(config,|_, _| ())?;
435:                            Ok(sc_service_test::TestNetComponents::new(keep_alive, rpc_handlers, client, network, transaction_pool))
438:                            let (keep_alive, rpc_handlers, client, network, transaction_pool) = new_light_base(config)?;
439:                            Ok(sc_service_test::TestNetComponents::new(keep_alive, rpc_handlers, client, network, transaction_pool))

bin/node/cli/src/service.rs
190:            prometheus_registry, telemetry_on_connect_sinks, rpc_handlers, ..
320:    Ok((task_manager, inherent_data_providers, rpc_handlers, client, network, transaction_pool))
326:    new_full_base(config, |_, _| ()).map(|(mut task_manager, rpc_handlers, _, _, _, _)| {
327:            task_manager.keep_alive(rpc_handlers);
343:            task_manager, rpc_handlers, client, network, transaction_pool, ..
428:    Ok((task_manager, rpc_handlers, client, network, transaction_pool))
433:    new_light_base(config).map(|(mut task_manager, rpc_handlers, _, _, _)| {
434:            task_manager.keep_alive(rpc_handlers);
496:                            let (keep_alive, inherent_data_providers, rpc_handlers, client, network, transaction_pool) = new_full_base(config, |
503:                            let node = sc_service_test::TestNetComponents::new(keep_alive, rpc_handlers, client, network, transaction_pool);
507:                            let (keep_alive, rpc_handlers, client, network, transaction_pool) = new_light_base(config)?;
508:                            Ok(sc_service_test::TestNetComponents::new(keep_alive, rpc_handlers, client, network, transaction_pool))
660:                            let (keep_alive, _, rpc_handlers, client, network, transaction_pool) = new_full_base(config, |_, _| ())?;
661:                            Ok(sc_service_test::TestNetComponents::new(keep_alive, rpc_handlers, client, network, transaction_pool))
664:                            let (keep_alive, rpc_handlers, client, network, transaction_pool) = new_light_base(config)?;
665:                            Ok(sc_service_test::TestNetComponents::new(keep_alive, rpc_handlers, client, network, transaction_pool))

client/service/test/src/lib.rs
83:     rpc_handlers: Arc<RpcHandlers>,
93:             rpc_handlers: RpcHandlers,
100:                    rpc_handlers: Arc::new(rpc_handlers),
115:                    rpc_handlers: self.rpc_handlers.clone()
[0] [14:44:18] ~/r/substrate ashley-service-multiple-return > rg rpc_handlers
utils/browser/src/lib.rs
119:pub fn start_client(mut task_manager: TaskManager, rpc_handlers: RpcHandlers) -> Client {
130:                                    let fut = rpc_handlers

client/service/src/builder.rs
1098:           let rpc_handlers = RpcHandlers(gen_handler(sc_rpc::DenyUnsafe::No));
1143:                   rpc_handlers,

client/service/src/lib.rs
169:    pub rpc_handlers: RpcHandlers,

bin/node/cli/src/browser.rs
55:     let (task_manager, rpc_handlers) =
57:                     .map(|(components, rpc_handlers, _, _, _)| (components, rpc_handlers))
60:     Ok(browser_utils::start_client(task_manager, rpc_handlers))

bin/node/cli/src/chain_spec.rs
434:                            let (keep_alive, _, rpc_handlers, client, network, transaction_pool) = new_full_base(config,|_, _| ())?;
435:                            Ok(sc_service_test::TestNetComponents::new(keep_alive, rpc_handlers, client, network, transaction_pool))
438:                            let (keep_alive, rpc_handlers, client, network, transaction_pool) = new_light_base(config)?;
439:                            Ok(sc_service_test::TestNetComponents::new(keep_alive, rpc_handlers, client, network, transaction_pool))

bin/node/cli/src/service.rs
190:            prometheus_registry, telemetry_on_connect_sinks, rpc_handlers, ..
320:    Ok((task_manager, inherent_data_providers, rpc_handlers, client, network, transaction_pool))
326:    new_full_base(config, |_, _| ()).map(|(mut task_manager, rpc_handlers, _, _, _, _)| {
327:            task_manager.keep_alive(rpc_handlers);
343:            task_manager, rpc_handlers, client, network, transaction_pool, ..
428:    Ok((task_manager, rpc_handlers, client, network, transaction_pool))
433:    new_light_base(config).map(|(mut task_manager, rpc_handlers, _, _, _)| {
434:            task_manager.keep_alive(rpc_handlers);
496:                            let (keep_alive, inherent_data_providers, rpc_handlers, client, network, transaction_pool) = new_full_base(config, |
503:                            let node = sc_service_test::TestNetComponents::new(keep_alive, rpc_handlers, client, network, transaction_pool);
507:                            let (keep_alive, rpc_handlers, client, network, transaction_pool) = new_light_base(config)?;
508:                            Ok(sc_service_test::TestNetComponents::new(keep_alive, rpc_handlers, client, network, transaction_pool))
660:                            let (keep_alive, _, rpc_handlers, client, network, transaction_pool) = new_full_base(config, |_, _| ())?;
661:                            Ok(sc_service_test::TestNetComponents::new(keep_alive, rpc_handlers, client, network, transaction_pool))
664:                            let (keep_alive, rpc_handlers, client, network, transaction_pool) = new_light_base(config)?;
665:                            Ok(sc_service_test::TestNetComponents::new(keep_alive, rpc_handlers, client, network, transaction_pool))

client/service/test/src/lib.rs
83:     rpc_handlers: Arc<RpcHandlers>,
93:             rpc_handlers: RpcHandlers,
100:                    rpc_handlers: Arc::new(rpc_handlers),
115:                    rpc_handlers: self.rpc_handlers.clone()
```

If we remove it from the return values, a lot of its occurrences will disappear.

Reply:
>:thinking: looks like the task manager has become our new "service", our node instance.

This is a bit concerning, but we can always split the task manager up again later on. And at least it's not very generic.

Reply:
Also, we can only create the rpc handlers after the task manager has been made, so it's not easy to put them together.

Reply:
I just realized that if it was using the keep_alive instead of being its own field it would be cool because you can add it in the task_manager after the task manager  is created.

But then again we might need it (browser::start_client).

SO: maybe the best option would be to:
1. create the task_manager
2. create the rpc_handlers
3. wrap rpc_handlers in Arc
4. clone it, and give it to keep_alive
5. return the other arc reference

Reply:
so if nobody needs it: it can be discarded in the return type safely

Reply:
(and also it makes it more "separate" from the task manager! :bulb: )